### PR TITLE
[WIP] Map interface

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: include/aos/common/tools/map.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:2]: 2024

File: include/aos/common/tools/map.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     * @return size_t.
+     */
+    size_t Size() const;
+
+    /**
+     * Returns maximum  allowed number of elements in the map.
+     *
+     * @return size_t.
+     */
+    size_t Capacity() const;
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:72]: MaxSize - to be aligned with aos::Array

File: include/aos/common/tools/map.hpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    Pair<Key,Value>*       end();
+    const Pair<Key,Value>* begin() const;
+    const Pair<Key,Value>* end() const;
+
+protected:
+    Map(Array<Pair<Key,Value>>& array);
+};
+
+}
+
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:87]: What about operator[] ?

File: include/aos/common/tools/map.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+template<typename Key, typename Value>
+class Map {
+public:
+    /**
+     * Returns pointer to the value for a given key. And null if a key is absent.
+     * 
+     * @param key
+     * @return const Value*
+     */
+    Value* Get(const Key& key);
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:28]: Can we use `At` that returns RetWithError<Value&>?

File: include/aos/common/tools/map.hpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Returns pointer to the value for a given key. And null if a key is absent.
+     * 
+     * @param key
+     * @return const Value*
+     */
+    const Value* Get(const Key& key) const;
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:36]: Ditto

File: include/aos/common/tools/map.hpp (hunk 4 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp:28]: I would leave name "Get" to have a pair with Set function. And return a pointer as it doesn't require additional variable in client code,  a bit easier to use.

File: include/aos/common/tools/map.hpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp:87]: operator[] inserts key if it is absent & for our case should return  RetWithError<Value&> or Value*.  This makes its usage different from builtin operators, that looks weird.
If required I can add Insert function which behaves the same.

File: include/aos/common/tools/map.hpp (hunk 2 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp]: fixed

File: include/aos/common/tools/map.hpp (hunk 1 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp]: fixed

File: include/aos/common/tools/map.hpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    RetWithError<const Value&> At(const Key& key) const;
+
+    /**
+     * Replaces map with elements from the list.
+     *
+     * @param list source list
+     * @return Error
+     */
+    Error Assign(const std::initializer_list<Pair<Key, Value>>& list);
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:44]: Unfortunatelly no std in `aos_core_lib_cpp`. We can replace this method with one:

` Error Assign(const Array<Pair<Key, Value>> list);` to assign from array.

Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp]: fixed

File: include/aos/common/tools/map.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    Error Set(const Key& key, const Value& value);
+
+    /**
+     * Removes element with the specified key from the map.
+     * 
+     * @param key
+     * @return Error
+     */
+    Error Erase(const Key& key);
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp]: fixed

File: include/aos/common/tools/map.hpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp]: fixed

File: include/aos/common/tools/utils.hpp [hunk 8]
```diff
@@ -44,6 +44,9 @@ struct Pair {
         : mFirst(f)
         , mSecond(s) {};
 
+    bool operator==(const Pair<F, S>& other) { return mFirst == other.mFirst && mSecond == other.mSecond; }
```
Comment (al1img) [inline include/aos/common/tools/utils.hpp:47]: Add comments for public items

File: include/aos/common/tools/map.hpp [hunk 9]
```diff
@@ -25,31 +29,61 @@ class Map {
...
      *
      * @param array source array
      * @return Error
      */
-    Error Assign(const Array<Pair<Key, Value>>& array);
+    Error Assign(const Array<Pair<Key, Value>>& array)
+    {
+        mItems.Clear();
+
+        for (const auto& [key, value] : array) {
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:61]: `mItems = array` ?

File: include/aos/common/tools/map.hpp [hunk 10]
```diff
@@ -25,31 +29,61 @@ class Map {
      * @param key
      * @return RetWithError<Value&>
      */
-    RetWithError<Value&> At(const Key& key);
+    RetWithError<Value&> At(const Key& key)
+    {
+        auto item = mItems.Find([&key](const Pair<Key, Value>& item) { return item.mFirst == key; });
+
+        if (!item.mError.IsNone()) {
+            return {*static_cast<Value*>(nullptr), item.mError};
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:37]: I have cppcheck warning here: `Null pointer dereference: static_cast<Value*>(nullptr)CppCheck (c-cpp-flylint)(nullPointer)`, suppress cppcheck.

File: include/aos/common/tools/map.hpp [hunk 11]
```diff
@@ -58,61 +92,114 @@ class Map {
...
-    const Pair<Key, Value>* begin() const;
-    const Pair<Key, Value>* end() const;
+    Pair<Key, Value>*       begin() { return mItems.begin(); }
+    Pair<Key, Value>*       end() { return mItems.end(); }
+    const Pair<Key, Value>* begin() const { return mItems.begin(); }
+    const Pair<Key, Value>* end() const { return mItems.end(); }
 
 protected:
-    Map(Array<Pair<Key, Value>>& array);
+    Map(Array<Pair<Key, Value>>& array)
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:177]: const Array?

File: include/aos/common/tools/map.hpp [hunk 12]
```diff
@@ -25,31 +29,61 @@ class Map {
...
+    RetWithError<const Value&> At(const Key& key) const { return const_cast<Map<Key, Value>&>(*this).At(key); }
 
     /**
      * Replaces map with elements from the array.
      *
      * @param array source array
      * @return Error
      */
-    Error Assign(const Array<Pair<Key, Value>>& array);
+    Error Assign(const Array<Pair<Key, Value>>& array)
```
Comment (al1img) [inline include/aos/common/tools/map.hpp:57]: Would be nice to have Emplace method.

File: tests/common/src/tools/map_test.cpp [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    StaticMap<std::string, int, 4> map;
+
+    EXPECT_TRUE(map.Assign(ConvertToArray(source)).IsNone());
+
+    // Remove existing key
+    EXPECT_TRUE(map.Remove("0xA").IsNone());
+    EXPECT_FALSE(map.At("0xF").mError.IsNone());
+
+    // Remove missing key: array doesn't return NotFound error. Map follows this behaviour.
+    // EXPECT_FALSE(map.Remove("0xD").IsNone());
```
Comment (al1img) [inline tests/common/src/tools/map_test.cpp:134]: Uncomment?

File: include/aos/common/tools/map.hpp (hunk 9 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/map.hpp:61]: doesnt handle duplicates

**Pull request merged**